src/templates/hl7v2/OML_O21.hbs

```diff
@@ -3,86 +3,110 @@
+
+            {{#if PID-3-1}},
+                {{>Resources/Patient.hbs PID=PID PD1=PD1 NK1=NK1 ID=(generateUUID PID-3-1)}},
+            {{/if}},
+
+            {{#if PV1}},
+
+                {{>Resources/Encounter.hbs PV1=PV1 PV2=PV2 ID=(generateUUID PV1)}},
+
+                {{#if PID-3-1}},
+                    {{>References/Encounter/subject.hbs ID=(generateUUID PV1) REF=(concat 'Patient/' (generateUUID PID-3-1))}},
+                {{/if}},
+
+            {{/if}},
+
             {{#with (getSegmentLists ../msg.v2 'AL1')}}
                 {{#each AL1 as |AL1Instance|}}
-    
+
                     {{>Resources/AllergyIntolerance.hbs AL1 ID=(generateUUID AL1Instance)}},
-                    {{>References/AllergyIntolerance/patient.hbs ID=(generateUUID AL1Instance) REF=(concat 'Patient/' (generateUUID ../../PID-3-1))}},
-    
+
+                    {{#if ../../PID-3-1}},
+                        {{>References/AllergyIntolerance/patient.hbs ID=(generateUUID AL1Instance) REF=(concat 'Patient/' (generateUUID ../../PID-3-1))}},
+                    {{/if}},
+
                 {{/each}}
             {{/with}} 	
             
             {{#with (getSegmentLists ../msg.v2 'NK1')}}
                 {{#each NK1 as |NK1Instance|}}
-    
+
                     {{>Resources/RelatedPerson.hbs NK1=NK1Instance ID=(generateUUID NK1Instance)}},
-                    {{>References/RelatedPerson/patient.hbs ID=(generateUUID NK1Instance) REF=(concat 'Patient/' (generateUUID ../../PID-3-1))}},
-    
+
+                    {{#if ../../PID-3-1}},
+                        {{>References/RelatedPerson/patient.hbs ID=(generateUUID NK1Instance) REF=(concat 'Patient/' (generateUUID ../../PID-3-1))}},
+                    {{/if}},
+
                 {{/each}}
             {{/with}} 
         
             {{#with (getSegmentLists ../msg.v2 'ORC')}}
                 {{#each ORC as |ORCInstance|}}
-                    
+
                     {{>Resources/ServiceRequest.hbs ORC=ORCInstance ID=(generateUUID (concat "ServiceRequest" ORCInstance))}},
                     {{>Resources/Provenance.hbs ORC=ORCInstance ID=(generateUUID (concat "Provenance" ORCInstance))}},
-                    {{>References/ServiceRequest/subject.hbs ID=(generateUUID (concat "ServiceRequest" ORCInstance)) REF=(concat 'Patient/' (generateUUID  ../../PID-3-1))}}
                     {{>References/Provenance/target.hbs ID=(generateUUID (concat "Provenance" ORCInstance)) REF=(concat 'ServiceRequest/' (generateUUID (concat "ServiceRequest" ORCInstance)))}},
-    
+
+                    {{#if ../../PID-3-1}},
+                        {{>References/ServiceRequest/subject.hbs ID=(generateUUID (concat "ServiceRequest" ORCInstance)) REF=(concat 'Patient/' (generateUUID  ../../PID-3-1))}}
+                    {{/if}},
+
                     {{#with (getRelatedSegmentList ../../../msg.v2 'ORC' ORCInstance-1 'TQ1')}}
                         {{#each TQ1 as |TQ1Instance|}}
-                            
+
```

Comment:
Same here for "getRelatedSegmentList". ("getSegmentLists", "getRelatedSegmentList" and "getFirstSegments" would better keep the same behavior.)

Reply:
Same reason with "ADT_A01".

Reply:
getRelatedSegmentLists seems a bit different from getFirstSegments. It returns a list instead of objects, can we ensure the object in returned list is not null or empty (just like getSegmentLists)?

Reply:
Sure we should also update it, I got a missunderstanding before .